File: leetcode/896.rs
/*
Question:
An array is monotonic if it is either monotone increasing or monotone decreasing.
An array nums is monotone increasing if for all i <= j, nums[i] <= nums[j]. An array nums is monotone decreasing if for all i <= j, nums[i] >= nums[j].
Given an integer array nums, return true if the given array is monotonic, or false otherwise.
*/

// Approach:
// 1) We find out the tone of the array first by checking if the difference between the first elements is negetive or positive
// 2) Then we loop through the array, checking the sign of the difference of adjacent elements 
//    and if at any point it is not equal to tone then we return false
// 3) Otherwise we return true at the end

impl Solution {
    pub fn is_monotonic(nums: Vec<i32>) -> bool {
        let mut tone = nums.last().unwrap() - nums[0];
        tone = if tone >= 0 {1} else {-1};
        for i in 1..nums.len() {
           let mut sign = nums[i]-nums[i-1];
           if sign == 0 {continue;}
           sign = sign/sign.abs();
           if sign != tone {return false;}
        }
        true
    }
}

File: leetcode/896.java
/*
Question:
An array is monotonic if it is either monotone increasing or monotone decreasing.
An array nums is monotone increasing if for all i <= j, nums[i] <= nums[j]. An array nums is monotone decreasing if for all i <= j, nums[i] >= nums[j].
Given an integer array nums, return true if the given array is monotonic, or false otherwise.
*/

// Approach:
// 1) If the array has less than 2 elements, return true
// 2) Create a variable delta and set to tot he difference of the first and second element
// 3) Loop through the rest of the elements in the array
//      If delta is negative and the current element is less than the last, return false
//      Otherwise if delta is positive and current element is more than the last, return false
//      Otherwise if delta is still 0, set it to the difference of the last and current element
// 4) Return true

class Solution {
    public boolean isMonotonic(int[] nums) {
        if(nums.length < 2) return true;
        int delta = nums[0] - nums[1];
        for(int i=1; i<nums.length; i++) {
            if(delta < 0 && nums[i] < nums[i-1])return false;
            else if(delta > 0 && nums[i] > nums[i-1])return false;
            else if(delta == 0) delta = nums[i-1] - nums[i];
        }
        return true;
    }
}
